Add filler for SparseLengthsWeightedSum

yinghai opened the pull request:
This diff adds support to fillers for SparseLengthsWeight* ops. It does 3 things:
- Add the fillers for SparseLengthsWeight* ops
- Add filling heuristics to consider the path of LengthsRangeFill -> Gather -> SparseLengthsWeightedSum, where the length input is shared by LengthsRangeFill and SparseLengthsWeightedSum. Therefore, we need to carefully bound the value of that length input so that at Gather, it does not index out-of-bound for the weight input of Gather.
- Fix and simplify the logic of math::RandFixedSum, where we just keep rejecting the generated value if it violates the invariants.

Differential Revision: D13048216

highker commented on caffe2/core/operator_schema.h:
```diff
@@ -371,15 +371,18 @@ class CAFFE2_API OpSchema {
       size_t key_index,
-      size_t length_index);
+      size_t length_index,
+      size_t weight_index = -1);
```
Not a big fan of this interface....I kinda missed `-1` + `size_t` until I see `if (weight_index < fillers->size())`. Maybe just leave two separate helpers so we don't have to have an extra `if` in the implementation.

yinghai replied:
Yeah, this diff is just testing a few things. I'd have int as input which is the case in C2.

highker commented on caffe2/core/operator_schema.cc:
```diff
@@ -352,24 +353,33 @@ static void SparseSegmentsFillerHelper(
 //            \_____/  \________/  \__/
 // lengths =    [3,        4,       2]
-OpSchema& OpSchema::ValueKeyLengthInputFillers(
+OpSchema& OpSchema::WeightedValueKeyLengthInputFillers(
```
If we separate the interfaces, `WeightedValueKeyLengthInputFillers` would be:
```lang=c++
WeightedValueKeyLengthInputFillers(...) {
  ValueKeyLengthInputFillers(...);
  // add assertion on weight_index if necessary
  (*fillers)[weight_index]
        .Min(0)
        .Max(shapes[weight_index].front() - 1)
        .Dist(FD_UNIFORM);
} 
```

yinghai replied:
This doesn't work because ValueKeyLengthInputFillers sets a function. We need to go into the function creation phase.

highker commented on caffe2/utils/filler.h:
```diff
@@ -41,8 +43,17 @@ class TensorFiller {
         CAFFE_ENFORCE_GE(max * tensor->numel(), fixed_sum);
-        math::RandFixedSum<Type, Context>(
-            tensor->numel(), min, max, fixed_sum_, data, context);
+        auto average = fixed_sum / tensor->numel();
```
Either we could have a new distribution in `math_cpu.cc` or incorporate this logic into `RandFixedSum`? Ideally we would avoid making these branches more complicated. (Check the TODO "replace filler distribution enum with a better abstraction" above).

yinghai replied:
This is pretty hacky and not working for all the cases. I just want to see how much we can get while keeping the current memoryless filler implementation. I'm getting less and less confident...

highker commented on caffe2/utils/math_cpu.cc:
```diff
@@ -2321,37 +2321,41 @@ CAFFE2_RAND_UNIFORM_INT(uint64_t);
+      double stdev = std::min(mean - a, b - mean);                          \
+      std::normal_distribution<double> distribution{mean, stdev / 4.0};     \
+      T value = distribution(context->RandGenerator());                     \
+      T remaining_sum_test = remaining_sum - value;                         \
```
Feel free to implement the real random fixed sum alg (https://www.mathworks.com/matlabcentral/fileexchange/9700-random-vectors-with-fixed-sum) lol. I tried a bit and gave up.

yinghai commented:
@highker I cleaned up things and fixed the interface. Now things are running so it's up for a review. :)

highker commented on caffe2/core/operator_schema.cc:
```diff
@@ -375,6 +387,31 @@ OpSchema& OpSchema::ValueKeyLengthInputFillers(
+    SparseLengthsFillerHelper(shapes, key_index, length_index, &fillers);
+    // fill in the keys (value_index is used to get the correct shape)
+    SparseSegmentsFillerHelper(shapes, value_index, key_index, &fillers);
+    // fill in the weigths
```
typo `weigths `

highker commented on caffe2/core/operator_schema.cc:
```diff
@@ -340,7 +341,17 @@ static void SparseLengthsFillerHelper(
 }
 
-static void SparseSegmentsFillerHelper(
+void SparseWeightssFillerHelper(
```
typo `Weightss`

highker commented on caffe2/utils/math_test.cc:
```diff
@@ -716,6 +716,23 @@ TEST_F(BroadcastTest, BroadcastFloatTest) {
+  std::vector<int> l(20);
+  math::RandFixedSum<int, CPUContext>(
+      20, 1, 1000, 1000, l.data(), cpu_context_.get());
+  for (const auto i : l) {
```
Assume this should be an assertion rather than logging?

yinghai replied:
Yeah, I need to remove this. I don't need to add any assertion though because `math::RandFixedSum` also will check the invariants at the end.

highker reviewed (approved):
Thanks for handling the "weighted" case!

Commit:
Add filler for SparseLengthsWeightedSum (#13949)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/13949

This diff adds support to fillers for `SparseLengthsWeight*` ops. It does 3 things:
1. Add the fillers for `SparseLengthsWeight*` ops
2. Add filling heuristics to consider the path of `LengthsRangeFill` -> `Gather` -> `SparseLengthsWeightedSum`, where the length input is shared by `LengthsRangeFill` and `SparseLengthsWeightedSum`. Therefore, we need to carefully bound the value of that length input so that at `Gather`, it does not index out-of-bound for the weight input of `Gather`.
3. Fix and simplify the logic of `math::RandFixedSum`, where we just keep rejecting the generated value if it violates the invariants.

Reviewed By: highker

Differential Revision: D13048216

fbshipit-source-id: 4410117798fa8be3e2af0ccfee48f8a137778296
